ton_client/src/net/websocket_link.rs

```diff
@@ -376,6 +399,32 @@ impl LinkHandler {
         next_phase
     }
 
+    async fn check_latency(&mut self) -> Option<Phase> {
+        let current = if let Some(x) = self.state.query_endpoint().await {
+            x
+        } else {
+            return None;
+        };
+        let env = self.client_env.clone();
+        if env.now_ms() < current.next_latency_detection_time {
+            return None;
+        }
+        let resolved = Endpoint::resolve(env, &self.config, &current.query_url).await;
+        match resolved.map(|x| x.latency) {
```

Comment:
If resolving failed then broadcast some specific error instead of `Bad latency detected`.